src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1923,6 +1937,31 @@ AssignTasksToConnectionsOrWorkerPool(DistributedExecution *execution)
 }
 
 
+/*
+ * WorkerPoolCompare is based on WorkerNodeCompare function.
+ *
+ * The function compares two worker nodes by their host name and port
+ * number.
+ */
+static int
+WorkerPoolCompare(const void *lhsKey, const void *rhsKey)
+{
+	const WorkerPool *workerLhs = *(const WorkerPool **) lhsKey;
+	const WorkerPool *workerRhs = *(const WorkerPool **) rhsKey;
+
+	int nameCompare = strncmp(workerLhs->nodeName, workerRhs->nodeName,
```

Comment:
Maybe just sort by node ID?

Reply:
I remember that we removed  `FindWorkerNode` calls in the past, do we want to bring them  back? https://github.com/citusdata/citus/pull/2848/commits/a3811b1e5517b8c0c5ae3a25f8753c4d9768331b

It seems that the only `findWorkerNode` was the one that we called where we need now.

Reply:
I might do some benchmarks before deciding on this

Reply:
It turns out that calling `FindWorkerNode` is pretty expensive compared to this sorting
```
worker_name_length | worker count | sort by name + compare by name | sort by id + compare by id
-- | -- | -- | --
15 | 3 | 0 microseconds | 3 microseconds
15 | 5 | 1 microseconds | 3 microseconds
15 | 10 | 2 microseconds | 4 microseconds
15 | 20 | 2 microseconds | 20 microseconds
15 | 50 | 5 microseconds | 60 microseconds
15 | 150 | 15 microseconds | 228 microseconds
15 | 500 | 62 microseconds | 755 microseconds
50 | 3 | 0 microseconds | 3 microseconds
50 | 5 | 1 microseconds | 3 microseconds
50 | 10 | 2 microseconds | 4 microseconds
50 | 20 | 2 microseconds | 20 microseconds
50 | 50 | 5 microseconds | 60 microseconds
50 | 150 | 16 microseconds | 228 microseconds
50 | 500 | 58 microseconds | 755 microseconds
200 | 3 | 0 microseconds | 3 microseconds
200 | 5 | 1 microseconds | 3 microseconds
200 | 10 | 2 microseconds | 4 microseconds
200 | 20 | 2 microseconds | 20 microseconds
200 | 50 | 4 microseconds | 60 microseconds
200 | 150 | 14 microseconds | 228 microseconds
200 | 500 | 64 microseconds | 755 microseconds
```

worker_name_length: means only the last characters differ in the workerNames, so strncmp goes to the end

Reply:
I guess we would have to add the node ID to the shard placement, but perhaps that's probably overengineering this a bit.